policykit/integrations/slack/models.py

```diff
@@ -176,41 +174,107 @@ def handle_metagov_event(self, outer_event):
+        this SlackCommunity gets updated (e.g. if a vote was cast).
+        """
+        assert process["name"] == "slack.emoji-vote"
+        outcome = process["outcome"]
+        status = process["status"]  # TODO: handle 'completed' status, which means that process was "closed"
+        ts = outcome["message_ts"]
+        votes = outcome["votes"]
+
+        action = PlatformAction.objects.filter(community=self, community_post=ts).first()
+        action_bundle = PlatformActionBundle.objects.filter(community=self, community_post=ts).first()
+        if action is not None:
+            # Expect this process to be a boolean vote on an action.
+            for (k, v) in votes.items():
+                assert k == "yes" or k == "no"
+                reaction_bool = True if k == "yes" else False
+                for u in v["users"]:
+                    user, _ = SlackUser.objects.get_or_create(username=u, community=self)
+                    existing_vote = BooleanVote.objects.filter(proposal=action.proposal, user=user).first()
+                    if existing_vote is None:
+                        logger.debug(f"Casting boolean vote {reaction_bool} by {user} for {action}")
+                        BooleanVote.objects.create(proposal=action.proposal, user=user, boolean_value=reaction_bool)
+                    elif existing_vote.boolean_value != reaction_bool:
+                        logger.debug(f"Casting boolean vote {reaction_bool} by {user} for {action} (vote changed)")
+                        existing_vote.boolean_value = reaction_bool
+                        existing_vote.save()
+
+        elif action_bundle is not None:
+            # Expect this process to be a choice vote on an action bundle.
+            bundled_actions = list(action_bundle.bundled_actions.all())
+            for (k, v) in votes.items():
+                num, voted_action = [(idx, a) for (idx, a) in enumerate(bundled_actions) if str(a) == k][0]
+                for u in v["users"]:
+                    user, _ = SlackUser.objects.get_or_create(username=u, community=self)
+                    existing_vote = NumberVote.objects.filter(proposal=voted_action.proposal, user=user).first()
+                    if existing_vote is None:
+                        logger.debug(
+                            f"Casting number vote {num} by {user} for {voted_action} in bundle {action_bundle}"
+                        )
+                        NumberVote.objects.create(proposal=voted_action.proposal, user=user, number_value=num)
+                    elif existing_vote.number_value != num:
+                        logger.debug(
+                            f"Casting number vote {num} by {user} for {voted_action} in bundle {action_bundle} (vote changed)"
+                        )
+                        existing_vote.number_value = num
+                        existing_vote.save()
+
+    def post_message(self, text, users=[], post_type="channel", channel=None):
```

Comment:
This function doesn't get called by anything, but, I created it just because I didn't want to throw away this code. The `im` and `ephemeral` post types didn't make sense to keep in `notify_action`, because you can't perform votes in those contexts. As discussed, we should separate out the authoring API "voting" from "notification," at which point we can re-use this code for notifying users.

Reply:
https://github.com/amyxzhang/policykit/issues/413

Reply:
Confused - why can't you perform votes in those contexts? I originally had those in (though not really fleshed out) because you might want private voting.

Clarifying my thoughts below - does this line up with what you are thinking?

Yes we should separate the act of sending a message to users in some way - which we want policy authors to be able to do whenever. You should be able to send any kind of message in private, public, mpim, ephemeral, etc. This should be like a "post_message()" function.

In the case of voting, there should be a function for enacting a voting, like "call_vote()", which then itself calls post_message() so that people have something to vote on. So settings for how to "notify" people about a vote need to be passed by the policy author through the voting function.

Reply:
> Confused - why can't you perform votes in those contexts? I originally had those in (though not really fleshed out) because you might want private voting.

You can't perform votes on `ephemeral` message because they don't support emoji-reactions. `im` voting makes sense, actually, you're right. It just wasn't implemented yet (there was/is no support for tallying votes on multiple `community_post`s). Agreed this should be a feature to add (to the Metagov Slack emoji-vote process)!

> Yes we should separate the act of sending a message to users in some way - which we want policy authors to be able to do whenever. You should be able to send any kind of message in private, public, mpim, ephemeral, etc. This should be like a "post_message()" function.

yes agreed

> In the case of voting, there should be a function for enacting a voting, like "call_vote()", which then itself calls post_message() so that people have something to vote on. So settings for how to "notify" people about a vote need to be passed by the policy author through the voting function.

Does it *need* to work like this internally? We are using Metagov, so, the fact that the voting mechanism uses the same Slack request as "post_message" internally is something that is hidden in the Metagov Slack plugin (and yes, [it currently does](https://github.com/metagov/metagov-prototype/blob/6d4506ec9ada78e3d538280d5f56d53bd7a077fe/metagov/metagov/plugins/slack/models.py#L210)).

The way I implemented, it will work more like this:
* VOTING: Policy author writes `slack_community.perform_vote(kind="mpim", users=["U1","U2","U3"])`. The `perform_vote` function makes a request to Metagov: `POST /api/internal/process/slack.emoji-vote {"message_type": "mpim", "users": [..], "text: "..." }`
* MESSAGING: Policy author writes `slack_community.post_message(kind="mpim", users=["U1","U2","U3"])`. The `post_message` function makes a request to Metagov: `POST /api/internal/action/slack.post-message {"kind": "mpim", "users": [..], "text": "..." }`

Reply:
Thanks for the pointer regarding ephemeral! I don't think we need it for voting then since you can provide that same functionality by letting a policy delete a message after some time/condition. I also think this is not high priority.

Well I don't have a strong opinion about WHERE it happens, in PK or MG, but I just wanted to point out that one will need to post a message when eliciting votes since emoji votes are only on messages (and right now in PK, we have to store which message ID so we know where to check for emoji votes). If you've got it covered in MG that sounds good.

Reply:
what about "initiate_vote()" or "initiate_vote_process()" or "initiate_voting()" instead of "perform_vote()"? Thinking about whether it could get confused with making an individual vote or carrying out a vote.

Reply:
👍 I like the "initiate_" options you mentioned. I won't make any changes to the authoring API without discussing with you & documenting them.